apps/emqx_dashboard/src/emqx_dashboard_admin.erl

```diff
@@ -378,44 +477,146 @@ all_users() ->
             #?ADMIN{
                 username = Username,
                 description = Desc,
-                role = Role
+                role = Role,
+                extra = Extra
             }
         ) ->
             flatten_username(#{
                 username => Username,
                 description => Desc,
-                role => ensure_role(Role)
+                role => ensure_role(Role),
+                mfa => format_mfa(Extra)
             })
         end,
         ets:tab2list(?ADMIN)
     ).
+
+format_mfa(#{mfa_state := #{mechanism := Mechanism}}) -> Mechanism;
+format_mfa(#{mfa_state := disabled}) -> disabled;
+format_mfa(_) -> none.
+
 -spec return({atomic | aborted, term()}) -> {ok, term()} | {error, Reason :: binary()}.
 return({atomic, Result}) ->
     {ok, Result};
 return({aborted, Reason}) ->
     {error, Reason}.
 
-check(undefined, _) ->
+check(Username, Password) ->
+    check(Username, Password, ?NO_MFA_TOKEN).
+
+check(undefined, _, _) ->
     {error, <<"username_not_provided">>};
-check(_, undefined) ->
+check(_, undefined, _) ->
     {error, <<"password_not_provided">>};
-check(Username, Password) ->
+check(Username, Password, MfaToken) ->
     case lookup_user(Username) of
         [#?ADMIN{pwdhash = PwdHash} = User] ->
-            case verify_hash(Password, PwdHash) of
-                ok -> {ok, User};
-                error -> {error, <<"password_error">>}
-            end;
+            IsPwdOk = (ok =:= verify_hash(Password, PwdHash)),
+            MfaVerifyResult = verify_mfa_token(Username, MfaToken, IsPwdOk),
+            check_mfa_pwd(MfaVerifyResult, IsPwdOk, User);
```

Comment:
I think that we shouldn't even do anything further once we establish that the password is incorrect.

i.e.:

```suggestion
            maybe
                true ?= (ok =:= verify_hash(Password, PwdHash)) || {error, <<"password_error"},
                MfaVerifyResult = verify_mfa_token(Username, MfaToken),
                check_mfa_pwd(MfaVerifyResult, User)
            end;
```

Reply:
For TOTP, we can (and should) return token error over password error.
Otherwise a brute-force attacker can send any token to guess the password like without MFA.

For other mechanisms which costs to send the token (e.g. email / SMS, if we will need to implement), password needs to be checked before token.